storage/integration_test.go

```diff
@@ -2573,259 +2576,348 @@ func TestIntegration_BucketIAM(t *testing.T) {
+				_, err = readObject(ctx, bucket.Object(objectName))
+				checkforErrors("read object", err)
+				_, err = bucket.Object(objectName).Attrs(ctx)
+				checkforErrors("get object attrs", err)
+				_, err = bucket.Object(objectName).Update(ctx, ObjectAttrsToUpdate{ContentLanguage: "en"})
+				checkforErrors("update object", err)
+
+				// Object ACL operations
+				checkforErrors("object acl set", bucket.Object(objectName).ACL().Set(ctx, entity, RoleReader))
+				_, err = bucket.Object(objectName).ACL().List(ctx)
+				checkforErrors("object acl list", err)
+				checkforErrors("object acl list", bucket.Object(objectName).ACL().Delete(ctx, entity))
+
+				// Default object ACL operations
+				// Once again, we interleave buckets to avoid rate limits
+				checkforErrors("default object acl set", bucket.DefaultObjectACL().Set(ctx, entity, RoleReader))
+				_, err = bucket.DefaultObjectACL().List(ctx)
+				checkforErrors("default object acl list", err)
+				checkforErrors("default object acl delete", bucket.DefaultObjectACL().Delete(ctx, entity))
+
+				// Copy
+				_, err = bucket.Object("copy").CopierFrom(bucket.Object(objectName)).Run(ctx)
+				checkforErrors("copy", err)
+				// Delete "copy" object, if created
+				if err == nil {
+					t.Cleanup(func() {
+						h.mustDeleteObject(bucket.Object("copy"))
+					})
+				}
 
+				// Compose
+				_, err = bucket.Object("compose").ComposerFrom(bucket.Object(objectName), bucket.Object("copy")).Run(ctx)
+				checkforErrors("compose", err)
+				// Delete "compose" object, if created
+				if err == nil {
+					t.Cleanup(func() {
+						h.mustDeleteObject(bucket.Object("compose"))
+					})
+				}
+
+				// Delete object
+				if err = bucket.Object(objectName).Delete(ctx); err != nil {
+					// We still want to delete object if the test errors
+					h.mustDeleteObject(requesterPaysBucket.Object(objectName))
+				}
+				checkforErrors("delete object", err)
+			})
+		}
+	})
+}
+
+// This test needs a second project and user to test all possibilities. Since we
+// need these things for Firestore already, we use them here.
+//
+// There are up to three entities involved in a requester-pays call:
+//  1. The user making the request. Here, we use the account used for the
+//     Firestore tests. The following must hold for this test to work:
+//     - this user must NOT have resourcemanager.projects.createBillingAssignment
+//     on the project that owns the bucket (2).
+//     - this user must have serviceusage.services.use permission on the Firestore
+//     project (3b).
+//     - this user must NOT have that serviceusage.services.use permission on
+//     the project that owns the bucket (3a).
+//  2. The project that owns the requester-pays bucket. Here, that
+//     is the test project ID (see testutil.ProjID).
+//  3. The project provided as the userProject parameter of the request;
+//     the project to be billed. This test uses:
+//     a. The project that owns the requester-pays bucket (same as (2))
```

Comment:
Is this true, or is it only 3b. that gets used in the non-owner test?

Also, you should add a description above the owner test as well. Nice rewrite but this is confusing stuff.